src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+    receiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle ackToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !ackToSend.IsNull());
+
+    // Pass BlockAck to sender and verify it was received
+    err = sender.HandleMessageReceived(ackToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    sender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_AckReceived);
+}
+
+// Test a full transfer using a responding receiver and an initiating sender, receiver drive.
+void TestInitiatingReceiverReceiverDrive(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TransferSession::OutputEvent outEvent;
+    TransferSession initiatingReceiver;
+    TransferSession respondingSender;
+
+    // Chosen arbitrarily for this test
+    uint32_t numBlockSends     = 4;
+    uint16_t transferBlockSize = 10;
+
+    uint32_t numBlocksSent = 0;
+
+    // Initialize struct with TransferInit parameters
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlagsRaw = kControl_ReceiverDrive;
+    initOptions.MaxBlockSize        = transferBlockSize;
+    char testFileDes[9]             = { "test.txt" };
+    initOptions.FileDesLength       = 9;
+    initOptions.FileDesignator      = reinterpret_cast<uint8_t *>(testFileDes);
+
+    // Verify receiver outputs ReceiveInit message after StartTransfer()
+    err = initiatingReceiver.StartTransfer(kRole_Receiver, initOptions);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    initiatingReceiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    NL_TEST_ASSERT(inSuite, !outEvent.MsgData.IsNull());
+
+    // Initialize respondingSender and pass ReceiveInit message
+    BitFlags<uint8_t, TransferControlFlags> senderOpts;
+    senderOpts.Set(kControl_ReceiverDrive);
+    err = respondingSender.WaitForTransfer(kRole_Sender, senderOpts, transferBlockSize);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    err = respondingSender.HandleMessageReceived(outEvent.MsgData.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Responder may want to read any of the fields of the TransferInit for approval.
+    // Verify File Designator matches. Verify that the respondingSenser automatically set ReceiverDrive
+    // for the chosen ControlMode.
+    respondingSender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_InitReceived);
+    NL_TEST_ASSERT(inSuite, respondingSender.GetControlMode() == kControl_ReceiverDrive);
+    NL_TEST_ASSERT(inSuite, outEvent.transferInitData.FileDesignator != nullptr);
+    NL_TEST_ASSERT(inSuite,
+                   !memcmp(testFileDes, outEvent.transferInitData.FileDesignator, outEvent.transferInitData.FileDesLength));
```

Comment:
This seems to also indicate the output message will include nul-terminator for strings, which it shouldn't since we are using prefix-length-based encoding for the file designator.